nixos/modules/services/web-servers/phpfpm/default.nix

```diff
@@ -48,6 +31,83 @@ let
     cat $phpPackage/etc/php.ini $nixDefaultsPath $phpOptionsPath > $out
   '';
 
+  poolOpts = { lib, name, ... }:
+    let
+      poolOpts = cfg.pools."${name}";
+    in
+    {
+      options = rec {
+        socket = mkOption {
+          type = types.str;
+          readOnly = true;
+          description = ''
+            Path to the unix socket file on which to accept FastCGI requests.
+            <note><para>This option is read-only and managed by NixOS.</note></para>
+          '';
+        };
+
+        listen = mkOption {
+          type = types.str;
+          default = "";
+          example = "/path/to/unix/socket";
+          description = ''
+            The address on which to accept FastCGI requests.
+          '';
+        };
+
+        phpPackage = mkOption {
+          type = types.package;
+          default = cfg.phpPackage;
+          defaultText = "config.services.phpfpm.phpPackage";
+          description = ''
+            The PHP package to use for running this PHP-FPM pool.
+          '';
+        };
+
+        phpOptions = mkOption {
+          type = types.lines;
+          default = cfg.phpOptions;
+          defaultText = "config.services.phpfpm.phpOptions";
+          description = ''
+            "Options appended to the PHP configuration file <filename>php.ini</filename> used for this PHP-FPM pool."
+          '';
+        };
+
+        user = mkOption {
+          type = types.str;
+          description = "User account under which this pool runs.";
+        };
+
+        group = mkOption {
+          type = types.str;
+          default = user;
+          description = "Group account under which this pool runs.";
+        };
+
+        extraConfig = mkOption {
+          type = types.lines;
+          example = ''
+            pm = dynamic
+            pm.max_children = 75
+            pm.start_servers = 10
+            pm.min_spare_servers = 5
+            pm.max_spare_servers = 20
+            pm.max_requests = 500
+          '';
+
+          description = ''
+            Extra lines that go into the pool configuration.
+            See the documentation on <literal>php-fpm.conf</literal> for
+            details on configuration directives.
+          '';
+        };
+      };
+
+      config = {
+        socket = if poolOpts.listen == "" then "${stateDir}/${name}.sock" else poolOpts.listen;
```

Comment:
Why not just mkDefault this for `listen` and drop the hidden `socket` option?

Reply:
The idea is to eventually eliminate the already existing `listen` option.

Reply:
I don't see the difference.. Also I'd expect the sockets to be created in `/run` and not in the `stateDir`.

Reply:
Yeah... they are created in `/run`. For some reason an author at some point decided to use `stateDir` as a variable name for `/run/phpfpm`. I guess I should rename that to make things less confusing while I'm at it.

The difference is that existing code using `listen` doesn't break and the user gets a warning to stop using `listen` in favor of `socket`. In a subsequent release (20.03) we can then entirely remove `listen` option and configuration using `phpfpm` will use the read only `socket` option to get the socket path (which is better than magic strings, of course).

I hope I have explained my intention sufficiently. Please let me know if further clarification is required.

Reply:
But what is the improvement we get by using `socket` vs. `listen`. I guess there are a few valid use cases for wanting to change that value, e.g. listening on a TCP socket. I'd prefer not having `default = ""`, but rather `nullOr str` and `default = null;` and then a `mkDefault "/run/php-fpm/${poolName}.sock"` which makes it accessible from other modules but still allows overriding. I don't see any further advantages we get by making it read-only and only the disadvantage mentioned above.

Reply:
One motivation I had in doing so was to provide a smooth transition from running `phpfpm` as `root` to non `root`.

If we keep the `listen` option it seems we can't easily continue to use `/run/phpfpm` as a directory to store sockets, but instead would need to switch to something like `/run/phpfpm/${poolName}` so that each individual `phpfpm` service can own its own directory for sockets. By keeping `listen` we either have to determine that `listen` is a socket (and not tcp) and then create the required directories (and potentially even the socket file itself to account for permissions) *or* tell the user if they are using a non standard directory they have to ensure the appropriate permissions themselves. Regardless this ends up in a situation where existing code can break without much warning to the user, unless we go through what I would consider an unmatched amount of effort in module code for little reward.

Or maybe I just haven't considered all the options for keeping things working with little inconvenience (and no possibility of runtime breakage) to the user :thinking:

Reply:
@globin ping

Reply:
@globin I've pushed some more changes which would allow me to drop the `socket` option and eventually make `listen` `readOnly = true;` in a future release - but still allowing the user to override the value via the new `settings` option. I'll hack this up tonight and hopefully it is more appealing to you.

Reply:
Pushed. Of course this means that we will get build time errors, but the build time errors combined with release notes hopefully aren't too disruptive. If anyone thinks they are, we can go back to a `socket` option (though I think `listen` is probably better as it is implemented now).

Reply:
Seeing the new code now makes me think the `socket` option I originally had was better. The `socket` option provides a much smoother transition - seamless, in fact.

Looking for weigh in from @globin @etu @Ma27 @wmertens and @Mic92 - are you ok with the previous version of the code which added a `socket` option and provided a smooth transition for users, or are the points @globin mentioned above enough to stick to `listen`?